Add instruction for custom istio gateway (#3802)
knative/docs: docs/serving/setting-up-custom-ingress-gateway.md

Knative service mesh, which is the `knative-ingress-gateway` Gateway under
the `knative-serving` namespace. By default, we use Istio gateway service
`istio-ingressgateway` under `istio-system` namespace as its underlying service.
You can replace the service and the gateway with that of your own as follows.

## Replace the default `istio-ingressgateway` service

> #### Step 1: Create Gateway Service and Deployment Instance

Reviewer: The style in the Knative docs is to use sentence case in headings.

If there is a change in service ports (compared with that of
`istio-ingressgateway`), update the port info in the gateway accordingly.

### Step 3: Update the gateway ConfigMap

1. Update gateway configmap `config-istio` under `knative-serving`
namespace:

     ```bash
     kubectl edit configmap config-istio -n knative-serving
     ```

     This command opens your default text editor and allows you to edit the config-istio ConfigMap.

     ```yaml
     apiVersion: v1
     data:
       _example: |
         ################################
         #                              #
         #    EXAMPLE CONFIGURATION     #
         #                              #
         ################################
         # ...
         gateway.knative-serving.knative-ingress-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
     ```

> 1. Edit the file to replace the `istio-ingressgateway.istio-system.svc.cluster.local` field with

Reviewer 2: I feel "replace" might be confusing as we actually "add" a new field into "data" section.

First, we create the `knative-custom-gateway` gateway.

```
cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: knative-custom-gateway
  namespace: custom-ns
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
EOF
```

!!! note
    Replace the label selector `istio: ingressgateway` with the label of your service.

### Step 2: Update the gateway ConfigMap

1. Update gateway configmap `config-istio` under `knative-serving`
namespace:

     ```bash
     kubectl edit configmap config-istio -n knative-serving
     ```

     This command opens your default text editor and allows you to edit the config-istio ConfigMap.

     ```yaml
     apiVersion: v1
     data:
       _example: |
         ################################
         #                              #
         #    EXAMPLE CONFIGURATION     #
         #                              #
         ################################
         # ...
         gateway.knative-serving.knative-ingress-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
     ```

> 1. Edit the file to replace the `gateway.knative-serving.knative-ingress-gateway` field with

Reviewer 2: same here. we may not want to use "replace".

```
gateway.knative-serving.knative-ingress-gateway: custom-ingressgateway.custom-ns.svc.cluster.local
```

> ## Replace the knative-ingress-gateway gateway

Reviewer: ```suggestion
## Replace the `knative-ingress-gateway` gateway
```

We customized the gateway service so far, but we may also want to use our own gateway.
We can replace the default gateway with our own gateway with following steps.

#### Step 1: Create Gateway

Let's say you replace the default `knative-ingress-gateway` gateway with
`knative-custom-gateway` in `custom-ns`.
First, we create the `knative-custom-gateway` gateway.

```
cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: knative-custom-gateway
  namespace: custom-ns
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
EOF
```

!!! note
    Replace the label selector `istio: ingressgateway` with the label of your service.

### Step 2: Update Gateway Configmap

> Update gateway configmap `config-istio` under `knative-serving`

Reviewer: ```suggestion
Update the gateway ConfigMap `config-istio` under the `knative-serving`
```

> namespace:

Reviewer: ```suggestion
namespace by running the command:
```

```bash
kubectl edit configmap config-istio -n knative-serving
```

Replace the `gateway.knative-serving.knative-ingress-gateway` field with
the customized gateway.

```
gateway.knative-serving.knative-ingress-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
```

For the gateway above, it should be updated to:

```
gateway.custom-ns.knative-custom-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
```

> The configuration format should be `gateway.GATEWAY_NAMESPACE.GATEWAY_NAME`.
> The `GATEWAY_NAMESPACE` is optional. when it is omitted, the system will search for

Reviewer: ```suggestion
The configuration format should be `gateway.<gateway-namespace>.<gateway-name>`.
`<gateway-namespace>` is optional. When it is omitted, the system searches for
```

Reviewer: Knative style rules say that placeholders should be in the format `<placeholder-name>` and not `PLACEHOLDER_NAME`.

> the gateway in the serving system namespace `knative-serving`.

Author: Sure! Let me try to change it.
